### PR TITLE
LLL: fix handling of "sha3" expression

### PR DESCRIPTION
**Pull request by benjaminion**
When PR #2317 changed the EVM opcode from SHA3 to KECCAK256 it broke the `(sha3 loc len)` expression in LLL. This PR fixes things while allowing existing code using the sha3 expression (such as the ENS registrar) to continue to compile.  I.e. both `(keccak256 loc len)` and `(sha3 loc len)` may be used, and the existing related sha3 macros continue to work.

### CODE REVIEW

**Comment by axic** [comment]
Can you please add an end to end test for both `keccak256` and `sha3`? That should have caught this in the first place 😉

File: liblll/CompilerState.cpp [hunk 1]
```diff
@@ -55,10 +55,10 @@ void CompilerState::populateStandard()
 	"(def 'msg (to data) { [0]:data (msg allgas to 0 0 32) })"
 	"(def 'create (value code) { [0]:(msize) (create value @0 (lll code @0)) })"
 	"(def 'create (code) { [0]:(msize) (create 0 @0 (lll code @0)) })"
+	"(def 'sha3 (loc len) (keccak256 loc len))"
 	"(def 'sha3 (val) { [0]:val (sha3 0 32) })"
```
Comment (axic) [inline liblll/CompilerState.cpp:60]: I think these two might be in conflict?

Comment (benjaminion) [reply liblll/CompilerState.cpp:60]: No, no conflict. The macro expansion takes into account the number of arguments so the two sha3 macros are distinguished correctly.

**Comment by benjaminion** [comment]
Three test cases added:

 - kekkac256
 - sha3 macro with 2 args
 - sha3 macro with 1 arg

File: test/liblll/EndToEndTest.cpp [hunk 2]
```diff
@@ -279,6 +279,43 @@ BOOST_AUTO_TEST_CASE(assembly_codecopy)
...
+{
+	char const* sourceCode = R"(
+		(returnlll
+			(seq
+				(mstore 0x00 0x01)
+				(return (keccak256 0x0 0x20))))
+	)";
+	compileAndRun(sourceCode);
+	BOOST_CHECK(callFallback() == encodeArgs(
+		fromHex("b10e2d527612073b26eecdfd717e6a320cf44b4afac2b0732d9fcbe2b7fa0cf6")));
```
Comment (axic) [inline test/liblll/EndToEndTest.cpp]: @chriseth can you please comment on the coding style you use for these kind of statements?

File: test/liblll/EndToEndTest.cpp [hunk 3]
```diff
@@ -279,6 +279,43 @@ BOOST_AUTO_TEST_CASE(assembly_codecopy)
 	BOOST_CHECK(callFallback() == encodeArgs(string("abcdef")));
 }
 
+BOOST_AUTO_TEST_CASE(keccak256_32bytes)
+{
+	char const* sourceCode = R"(
+		(returnlll
+			(seq
+				(mstore 0x00 0x01)
+				(return (keccak256 0x0 0x20))))
```
Comment (axic) [inline test/liblll/EndToEndTest.cpp]: Nitpicking: can you use `0x00`?

Comment (benjaminion) [reply test/liblll/EndToEndTest.cpp]: Fair comment! I've amended the commit accordingly.

**Comment by axic** [comment]
@benjaminion can you please rebase this? (Since the other LLL changes were merged this is in conflict.)

**Comment by benjaminion** [comment]
@axic 

Rebase done. Should be good - all LLL tests succeed.

**Comment by axic** [comment]
@benjaminion would you be interested adding tests for every internal macro in a separate PR? It would be nice catching these errors when we make changes 😉

**Review by axic — APPROVED** [review]

**Comment by axic** [comment]
Also thanks for this, merged!

**Comment by benjaminion** [comment]
@axic 

> would you be interested adding tests for every internal macro in a separate PR?

Sure. It'll be a week or two, but happy to take this one.

**Pull request merged**